Update commerce-sdk-react README (#1176)
SalesforceCommerceCloud/pwa-kit: packages/commerce-sdk-react/README.md

# Commerce SDK React

> <p align="center">

Reviewer: Non-Comment: Centre alignment is an interesting choice

Author: I feel like it's more eye catchy

A collection of React hooks for <b>fetching</b>, <b>caching</b>, and <b>mutating data</b> from the <b><a href="https://developer.salesforce.com/docs/commerce/commerce-api/overview">Salesforce B2C Commerce API</a></b> (SCAPI).

> The library contains declarative, always-up-to-date auto-managed <b>queries</b> and <b>mutations</b> for a more convenient developer experience.

Reviewer: Here are my thoughts on this section:

1. It's very market-y, and I don't really believe it reflects what the library actually is. We call it a "collection of React hooks" but I think we have previously settled that it's a collection of ReactQuery hooks which is more accurate.
2. Another part of our code base is the managing of authentication which we don't mention here.
3. We are also using a lot of ReactQuery terms like mutations/caching/queries as though it was something that we created. Instead I'd mentioned that we are taking advantage of these features that ReactQuery provides us.
4. "always-up-to-date auto-managed queries and mutations" I don't really know what that means. I think you might be referencing that is helps with upgradability with templates? which would be out of context.

Author: good feedback! Removed this confusing sentence and explained the tight integration with react query later in this doc

> The library contains declarative, always-up-to-date auto-managed <b>queries</b> and <b>mutations</b> that directly improve both developer and user experiences.

Reviewer 2: ```suggestion
The library contains declarative, always-up-to-date auto-managed <b>queries</b> and <b>mutations</b> for a more convenient developer experience.
```

</p>

## 🎯 Features

> - [react-query](https://tanstack.com/query/latest/docs/react/overview) and [commerce-sdk-isomorphic](https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic) integration

Reviewer: Would this be better suited for an architecture section or blurb in the intro. Aka is this really a feature?

Author: removed

- Shopper authentication & token management via [SLAS](https://developer.salesforce.com/docs/commerce/commerce-api/references/shopper-login)

> - SSR data fetching React Hooks

Reviewer: This is only true in the context of the PWA-Kit SDK's so is it really required here?

Reviewer: Maybe we can amend this to say that is works sever side in the context of pwa-kit apps

- Phased Launch support ([plugin_slas](https://github.com/SalesforceCommerceCloud/plugin_slas) compatible)

> - [TypeScript](https://www.typescriptlang.org/) Support

Reviewer: This again is an architectural detail what isn't really a feature. Maybe we can add it to the intro blur of add under an architectural section

> - Multi-site and locale/currency support

Reviewer: This is a feature of the API not necessarily this library. Its almost like saying listing getProduct as a feature.

- Request deduplication
- Built-in caching for easy state management

>   - auto cache invalidations/updates by mutations

Reviewer 2: ```suggestion
  - automatic cache invalidations/updates via the library's built-in mutations
```

>   - auto cache key generation

Reviewer 2: ```suggestion
  - automatic cache key generation
```

> ## ⚙️ Installation

Reviewer 3: Is adding a **Requirements** section with the Node engines requirements useful?

Are we missing a BSD 3-Clause license Information section?

Author: Hmmmmm, I don't believe node requirements is super important for this library. Unless someone ran into issues with a specific node version.

The license is [here](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/develop/LICENSE), do we need it in the README as well? @adamraya

```bash
npm install @salesforce/commerce-sdk-react @tanstack/react-query
```

## ⚡️ Quickstart (PWA Kit)

> To integration this library with your PWA Kit application you can use the `CommerceApiProvider` directly given that you use the `withReactQuery` higher order component to wrap your `AppConfig` component. Below is a snippet of how this is accomplished.

Reviewer 4: ```suggestion
To integrate this library with your PWA Kit application you can use the `CommerceApiProvider` directly, given that you use the `withReactQuery` higher order component to wrap your `AppConfig` component. Below is a snippet of how this is accomplished.
```

> To integration this library with your PWA Kit application you can use the `CommerceApiProvider` directly given that you use the `withReactQuery` higher order component to wrap your `AppConfig` component. Below is a snippet of how this is accomplished.

Reviewer 2: ```suggestion
To integrate this library with your PWA Kit application you can use the `CommerceApiProvider` directly assuming that you use the `withReactQuery` higher order component to wrap your `AppConfig` component. Below is a snippet of how this is accomplished.
```

> To integration this library with your PWA Kit application you can use the `CommerceApiProvider` directly given that you use the `withReactQuery` higher order component to wrap your `AppConfig` component. Below is a snippet of how this is accomplished.

Reviewer 2: ```suggestion
To integrate this library with your PWA Kit application you can use the `CommerceApiProvider` directly assuming that you use the `withReactQuery` higher order component to wrap your `AppConfig` component. Below is a snippet of how this is accomplished.
```

> To integrate this library with your PWA Kit application you can use the `CommerceApiProvider` directly, given that you use the `withReactQuery` higher order component to wrap your `AppConfig` component. Below is a snippet of how this is accomplished.

Author: ```suggestion
To integrate this library with your PWA Kit application you can use the `CommerceApiProvider` directly assuming that you use the `withReactQuery` higher order component to wrap your `AppConfig` component. Below is a snippet of how this is accomplished.
```

> To integrate this library with your PWA Kit application you can use the `CommerceApiProvider` directly assuming that you use the `withReactQuery` higher order component to wrap your `AppConfig` component. Below is a snippet of how this is accomplished.

Reviewer: Instead of leaking details of PWA-Kit by mentioning `withReactQuery` maybe we should simply specify the version number of the pwa kit that we know already have that HoC. 
In the future we can add another section for integration with older versions. I say this because we need to give those older sdk users server side support and we'd be using `getProps` to serialize the state of the provider.

Author: added pwa kit version.

I do think it's worth mentioning `withReactQuery` since the package is inside the monorepo and the primary use case is `withReactQuery`. This highlights the integration and gives more exposure of the SDK integration, otherwise developers might not know `withReactQuery` and things could be a bit black boxy for them. That's why i left the `withReactQuery` examples in the doc.

Reviewer: I'll meet you half way on this one 👍

export default withReactQuery(AppConfig, options)
```

## ⚡️ Quickstart (Generic React App)

> You can use this library in any React application provided you bring your own QueryClient and QueryClientProvider. Below is a sample integration:

Reviewer 5: ```suggestion
You can use this library in any React application by creating a new QueryClient and wrap your application with `QueryClientProvider`. For example: 
```

To help reduce boilerplate code for managing shopper authentication, by default, this library automatically initializes shopper session and manages the tokens for developers. Currently, the library supports the [Public Client login flow](https://developer.salesforce.com/docs/commerce/commerce-api/guide/slas-public-client.html).

### Shopper Session Initialization

On `CommerceApiProvider` mount, the provider initializes shopper session by initiating the SLAS __Public Client__ login flow. The tokens are stored/managed/refreshed by the library.

### Authenticate request queue

While the library is fetching/refreshing the access token, the network requests will queue up until there is a valid access token.

### Login helpers

To leverage the managed shopper authentication feature, use the `useAuthHelper` hook for shopper login.

Example:

```jsx
import {AuthHelpers, useAuthHelper} from '@salesforce/commerce-sdk-react'

const Example = () => {
    const register = useAuthHelper(AuthHelpers.Register)
    const login = useAuthHelper(AuthHelpers.LoginRegisteredUserB2C)
    const logout = useAuthHelper(AuthHelpers.LogOut)

    return <button onClick={() => {

>         login.mutate({username, password})

Reviewer: NIT: you might want to replace the variables username and password and provide example data. This will make it more obvious that if someone wants to copy and paste the code they have to change/provide that data, instead of ensuring there is some variables with those names defined somewhere.

Reviewer: also give you an opportunity to put you name on things ;)

```
 login.mutate({username: 'kevin', password: 'pa$$word'})
```

>         login.mutate({username, password})

Author: ```suggestion
        login.mutate({username: 'kevin', password: 'pa$$word'})
```

>         login.mutate({username, password})

Author: ```suggestion
        login.mutate({username: 'kevin', password: 'pa$$word'})
```

export default App
```

## Shopper Authentication and Token Management

_💡 This section assumes you have read and completed the [Authorization for Shopper APIs](https://developer.salesforce.com/docs/commerce/commerce-api/guide/authorization-for-shopper-apis.html) guide._

> To help reducing boilerplate code for managing shopper authentication, by default, this library automatically initializes shopper session and manage the tokens for developers. Currently, the library supports the [Public Client login flow](https://developer.salesforce.com/docs/commerce/commerce-api/guide/slas-public-client.html).

Reviewer 4: ```suggestion
To help reduce boilerplate code for managing shopper authentication, by default, this library automatically initializes shopper session and manages the tokens for developers. Currently, the library supports the [Public Client login flow](https://developer.salesforce.com/docs/commerce/commerce-api/guide/slas-public-client.html).
```

### Shopper Session Initialization

On `CommerceApiProvider` mount, the provider initializes shopper session by initiating the SLAS Public Client login flow. The tokens are stored/managed by the library to support Phased Launch use cases and multi-site setup.

> When the access token is expired, the library will automatically try to refresh the access token by calling the `/oauth2/token` endpoint with `grant_type=refresh_token`.

Reviewer: I feel like there is too many implementation details here. I think we need to shorten this up and only talk about what might be important to the end user. I think one of those important things is where tokens are stored and what names they are store under. And whether or not those names can be changed or not (I can't remember is that is something that is configurable or not).

Author: 🤔 The token names and how we store them **were**  important to users because the implementation was in user land and developers had to understand it to update the code and consume new changes. Now, with the hook library, the tokens also become implementation details.

Author: I think it's important to note that the hook lib uses **public client** instead of a private client at the moment.

Author: I'll remove the details for refresh_token stuff

> When the access token is expired, the library will automatically try to refresh the access token by calling the `/oauth2/token` endpoint with `grant_type=refresh_token`.

Reviewer 5: Do we want to provide a little information on where the session data is stored in? say, access token is stored in localeStorage while usid and cookies are stored in cookies

Author: 🤔 This is a really good question... I envision that developers don't need to know the details about where/how the tokens are stored by using this library. It **was** important information previously because the implmentation was in the user land and we've been making changes to them, so developers used to need to understand the details to consume changes.

### Authenticate request queue

While the library is getting/refreshing the access token, the network requests will queue up until there is a valid access token.

### Auth helpers

> To leverage the managed shopper authentication feature, you want to use the auth helper hooks for shopper login.

Reviewer 4: ```suggestion
To leverage the managed shopper authentication feature, use the auth helper hooks for shopper login.
```

Example:

```jsx
import {AuthHelpers, useAuthHelper} from '@salesforce/commerce-sdk-react'

const Example = () => {
    const register = useAuthHelper(AuthHelpers.Register)
    const login = useAuthHelper(AuthHelpers.LoginRegisteredUserB2C)
    const logout = useAuthHelper(AuthHelpers.LogOut)

    return <button onClick={() => {
        login.mutate({username, password})
    }}>
}
```

### Opt-out from managed shopper authentication

> To opt-out from the auto-login, request queue, and stop the library from managing tokens, pass `fetchedToken` (a SLAS JWT) to the provider.

Reviewer 4: ```suggestion
To opt-out of the auto-login, request queue, and token management features, pass `fetchedToken` (a SLAS JWT) to the provider.
```

```jsx
const MyComponent = ({children}) => {
    return (
        <CommerceApiProvider
            fetchedToken="xxxxxxxxxxxx"
        >
            {children}
        </CommerceApiProvider>
    )
} 

> ```

Reviewer: I'm not sure the title "Opt-out" is the best here. It almost makes it sound like "do you not want use to automatically authenticate on app load, then do this, but you can authenticate later" .. I think a better section title would be something that is implying the use of a external token or auth mechanism. Maybe something like this (poorly worded but you get the idea):

### Externally Managed Shopper Authentication

You have the option of manually authenticating by providing a shopper token as a prop. This is useful if you plan on using this lib in an application there the authentication mechanism is different.

> ## Queries

Reviewer: I feel like we have these part headings for big topics like "Queries" and "Mutations" .. but we don't have a segue into them. Do we need to add something here, like: 

"The query hooks library can be broken down into queries, things that get data, and mutations, things that change data. Below are the basics of using mutations and queries: "

Author: added segue for react-query intro

The query hooks correspond to the http GET endpoints from the SCAPI family. The query hooks follow the signature pattern:

```
use<EntityName>(CommerceClientOptions, ReactQueryOptions)
```

> For example, the both the __required__ and __optional__ parameters for the underlying `commerce-sdk-isomorphic` call is passed as the first parameter:

Reviewer 4: ```suggestion
For example, both the __required__ and __optional__ parameters for the underlying `commerce-sdk-isomorphic` call is passed as the first parameter:
```

```jsx
import {useProduct} from '@salesforce/commerce-sdk-react'

const Example = () => {
    const query = useProduct({
        parameters: {
            id: '25592770M',
            locale: 'en-US'
        }
    })

    return <>
        <p>isLoading: {query.isLoading}</p>
        <p>name: {query.data?.name}</p>
    </>
}
```

The second parameter is the react-query query options, for more detail, read [useQuery reference](https://tanstack.com/query/latest/docs/react/reference/useQuery).

```jsx
import {useBasket} from '@salesforce/commerce-sdk-react'

const onServer = typeof window === undefined

const Example = ({basketId}) => {
    const query = useBasket({
        parameters: {
            basketId: basketId
        },
    }, {
        // A common use case for `enabled` is
        // to conditionally fetch based on environment
        enabled: !onServer && basketId
    })
}
```

> ### Advanced: Query Key Structure

Reviewer: If the answer to the question "Is our key structure private?" is "Yes" then I would leave this entire section out. I feel its going to be difficult to give a complete picture of the entire cache structure in this doc, and those wanting to do advanced things will be using the chrome dev tools and they can sus it out from there. Maybe we need an example of using those dev tools?

Author: removed

> On most cases, you don't need to know this! If for advanced use cases where you want to manually update/invalidate the cache entires. Here is the explanation of the cache key structure.

Reviewer 4: ```suggestion
In most cases, you won't need to use this! This feature is intended for advanced use cases where you want to manually update/invalidate the cache entires. Here is the explanation of the cache key structure.
```

The query keys are automatically generated in the following format:

```js
['commerce-sdk-react', ...RestfulApiPathSegments, {...CommerceClientOptions}]
```

For example, `useProduct({parameters: {id: '25592770M', "allImages": true}})` hook has query key:

```js
['commerce-sdk-react', '/organizations', '/f_ecom_aaaa_001', '/products', '/25688443M', {
    "organizationId": "f_ecom_zzrf_001",
    "id": "25688443M",
    "currency": "GBP",
    "locale": "en-GB",
    "allImages": true,
    "siteId": "RefArchGlobal"
  } ]
```

> The first element in the query key is always a constant value `commerce-sdk-react`, the namespace provides flexibility for invalidating the entire query cache. This is typically useful when using react-query for multiple services including SCAPI. This allows us to invalidate the entire SCAPI cache without affecting other cache entries, for example, cache for fetching data from a customer CMS provider. To invalidate all SCAPI cache entries, use `queryClient.invalidateQueries(['commerce-sdk-react'])`.

Reviewer 4: ```suggestion
The first element in the query key is always a constant value, `commerce-sdk-react`. The namespace provides flexibility for invalidating the entire query cache. This is typically useful when using react-query for multiple services including SCAPI. This allows us to invalidate the entire SCAPI cache without affecting other cache entries, such as the cache for fetching data from a customer CMS provider. To invalidate all SCAPI cache entries, use `queryClient.invalidateQueries(['commerce-sdk-react'])`.
```

The following N elements in the array are the path segment of the actual SCAPI endpoint. This design makes it easier to have granular control over individual cache as well as the ability to do bulk invalidations based on the hierarchy of the entities.

Example, invalidate customer's basket v.s. invalidate just the shipping address in the basket

```js
// invalidate entire customer basket cache, including getBasket, getPaymentMethodsForBasket, getPriceBooksForBasket, getShippingMethodsForShipment and getTaxesFromBasket
queryClient.invalidateQueries(['commerce-sdk-react', '/organizations', '/f_ecom_aaaa_001', '/baskets', '/BASKET_ID'])

// granular invalidation for just the shipping methods
queryClient.invalidateQueries(['commerce-sdk-react', '/organizations', '/f_ecom_aaaa_001', '/baskets', '/BASKET_ID', '/shipments', 'SHIPMENT_ID', '/shipping-methods'])
```

> _💡 Hint: include `@tanstack/react-query-devtools` in your React app to see the queries (and their cache keys)._

Reviewer 5: ```suggestion
_💡 Hint: install and include `@tanstack/react-query-devtools` in your React app to see the queries (and their cache keys)._
```

> _💡 Hint: install and include `@tanstack/react-query-devtools` in your React app to see the queries (and their cache keys)._

Reviewer 3: nit: I think React Query Devtools is a really basic tool to work successfully with this library I'd highlight it a bit more, maybe in the setup steps as an optional step and then link it to the React Query Devtools docs: https://tanstack.com/query/latest/docs/react/devtools?

## Mutations

> The query hooks correspond to the http POST, PUT, PATCH, DELETE endpoints from the SCAPI family. The mutation hooks follow the signature pattern:

Reviewer 5: Nit: `SCAPI endpoints` or SCAPI backend/network sounds more formal than the word 'SCAPI family' 🤔

> The query hooks correspond to the http POST, PUT, PATCH, DELETE endpoints from the SCAPI family. The mutation hooks follow the signature pattern:

Reviewer 2: ```suggestion
The query hooks correspond to the http POST, PUT, PATCH, DELETE endpoints from the SCAPI API domains. The mutation hooks follow the signature pattern:
```

> The query hooks correspond to the http POST, PUT, PATCH, DELETE endpoints from the SCAPI family. The mutation hooks follow the signature pattern:

Reviewer 2: ```suggestion
The query hooks correspond to the http POST, PUT, PATCH, DELETE endpoints from the SCAPI API domains. The mutation hooks follow the signature pattern:
```

```

> use<ApiName>Mutations(EndpointName)

Reviewer 3: ```suggestion
use<ApiName>Mutation(EndpointName)
```

```

> For example, the [ShopperBaskets API](https://developer.salesforce.com/docs/commerce/commerce-api/references/shopper-baskets?meta=Summary) has a number of endpoints, one of the endpoint is the [addItemToBasket](https://developer.salesforce.com/docs/commerce/commerce-api/references/shopper-baskets?meta=addItemToBasket) endpoint (`POST /baskets/{basketId}/items`).

Reviewer 4: ```suggestion
For example, the [ShopperBaskets API](https://developer.salesforce.com/docs/commerce/commerce-api/references/shopper-baskets?meta=Summary) has a number of endpoints, one of them being the [addItemToBasket](https://developer.salesforce.com/docs/commerce/commerce-api/references/shopper-baskets?meta=addItemToBasket) endpoint (`POST /baskets/{basketId}/items`).
```

```jsx
import {useShopperBasketsMutation} from '@salesforce/commerce-sdk-react'

const Example = ({basketId}) => {
    // Typescript IDE intellisense for available options as string
    const addItemToBasket = useShopperBasketsMutation('addItemToBasket')

    return <button onClick={() => addItemToBasket.mutate({
        parameters: {
            basketId
        },
        body: {
            productId: '25592770M',
            price: 55,
            quantity: 1
        }
    })} />
}
```

You could also import the mutation options as a constant like:

```jsx
import {useShopperBasketsMutation, ShopperBasketsMutations} from '@salesforce/commerce-sdk-react'

const Example = ({basketId}) => {
    const addItemToBasket = useShopperBasketsMutation(ShopperBasketsMutations.AddItemToBasket)
    return ...
}
```

> ### Auto Cache Invalidations

Reviewer 3: We only talk about invalidations but we don't talk about **_update_** and **_remove_**.

Perhaps the example we use is not the best since we don't invalidate on the `addItemToBasket`, we only update.
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/4939034d7f1d5d9262d8061380e855c865ca7ef1/packages/commerce-sdk-react/src/hooks/ShopperBaskets/cache.ts#L93-L101

Author: added explanations for update remove and invalidate

Since mutations changes data on the server, the cache entries that are potentially affected by the mutation is automatically invalidated.

> For example, a `addItemToBasket` mutation will automatically invalidate `getBasket` query.

Reviewer 5: ```suggestion
For example, a `addItemToBasket` mutation will automatically invalidate `getBasket`  and `customerBaskets` query
```

> For example, a `addItemToBasket` mutation will automatically invalidate `getBasket`  and `customerBaskets` query

Reviewer 3: Should we use the query hooks names instead?

```suggestion
For example, a `addItemToBasket` mutation will automatically invalidate `useBasket`  and `useCustomerBaskets` queries.
```

> For example, a `addItemToBasket` mutation will automatically invalidate `useBasket`  and `useCustomerBaskets` queries.

Author: ```suggestion
For example, an `addItemToBasket` mutation automatically update `useBasket`  and `useCustomerBaskets` query cache, because the mutation result contains the information for the updated basket. In other cases, when the mutation response do not have the updated data, the library will invalidate the cache and trigger a re-fetch.
```

> For example, an `addItemToBasket` mutation automatically update `useBasket`  and `useCustomerBaskets` query cache, because the mutation result contains the information for the updated basket. In other cases, when the mutation response do not have the updated data, the library will invalidate the cache and trigger a re-fetch.

Author: ```suggestion
For example, an `addItemToBasket` mutation automatically update `useBasket`  and `useCustomerBaskets` query cache, because the mutation result contains the information for the updated basket. In other cases, when the mutation response do not have the updated data, the library will invalidate the cache and trigger a re-fetch. For the DELETE endpoints, the library removes the cache entries on successful mutations.
```

## Ultilities

Besides the collection of query hooks and mutation hooks, here are some ultility hooks to help you interact with SCAPI.

### `useCommerceApi()`

> This hook returns a set of SCAPI clients, which are already initialized client using the configurations passed to the provider. Note: this hook doesn't automatically include auth headers.

Reviewer 4: ```suggestion
This hook returns a set of SCAPI clients, which are already initialized using the configurations passed to the provider. Note: this hook doesn't automatically include auth headers.
```

```jsx
import {useCommerceApi, useAccessToken} from '@salesforce/commerce-sdk-react'

const Example = () => {
    const api = useCommerceApi()
    const {getTokenWhenReady} = useAccessToken()

    const fetchProducts = async () => {
        const token = await getTokenWhenReady()
        const products = await api.shopperProducts.getProducts({
                parameters: {ids: ids.join(',')},
                headers: {
                    Authorization: `Bearer ${token}`
                }
            })
        return products
    }

}
```

### `useAccessToken()`

```ts
useAccessToken() => {token: String, getTokenWhenReady: Promise}
```

This ultility hook give access to the managed SLAS access token.

> ### `useCustomerId()`

Reviewer 3: Are we also planning on publishing and liking the typescript auto-generated docs site in the readme.md file with the rest of the utils and functions?

Author: yes
